cli/enginecli/test_gofp.py

```diff
@@ -2604,8 +2976,334 @@ def test_player_is_rewarded_for_making_a_choice_in_stages_that_have_rewards(
+            stages,
+            True,
+            {"from": self.game_master},
+        )
+
+        session_id = self.gofp.num_sessions()
+
+        # Rewards for all paths except for stage 1 path 1 and stage 2 path 3.
+        stages_with_rewards = [1, 2, 2, 3]
+        paths_with_rewards = [2, 1, 2, 1]
+
+        self.gofp.set_path_rewards(
+            session_id,
+            stages_with_rewards,
+            paths_with_rewards,
+            [self.terminus.address] * len(stages_with_rewards),
+            [self.reward_pool_id] * len(stages_with_rewards),
+            [
+                (2**stage) * (3**path)
+                for stage, path in zip(stages_with_rewards, paths_with_rewards)
+            ],
+            {"from": self.game_master},
+        )
+
+        self.gofp.set_session_active(session_id, True, {"from": self.game_master})
+
+        # Mint NFTs to the player
+        total_nfts = self.nft.total_supply()
+
+        num_nfts = 8
+        token_ids = [total_nfts + i for i in range(1, num_nfts + 1)]
+
+        for token_id in token_ids:
+            self.nft.mint(self.player.address, token_id, {"from": self.owner})
+
+        # Mint num_tokens*payment_amount of payment_token to player
+        self.payment_token.mint(
+            self.player.address, len(token_ids) * payment_amount, {"from": self.owner}
+        )
+
+        self.payment_token.approve(self.gofp.address, MAX_UINT, {"from": self.player})
+        self.nft.set_approval_for_all(self.gofp.address, True, {"from": self.player})
+
+        self.gofp.stake_tokens_into_session(
+            session_id, token_ids, {"from": self.player}
+        )
+
+        # First half (rounded up) of tokens will choose path 1. Rest will choose path 2.
+        first_stage_path_1_count = math.ceil(num_nfts / 2)
+        first_stage_path_2_count = math.floor(num_nfts / 2)
+        first_stage_path_choices = [1] * first_stage_path_1_count + [
+            2
+        ] * first_stage_path_2_count
+
+        first_stage_tx_receipt = self.gofp.choose_current_stage_paths(
+            session_id,
+            token_ids,
+            first_stage_path_choices,
+            {"from": self.player},
+        )
+
+        first_stage_events = _fetch_events_chunk(
+            web3_client,
+            ERC1155_TRANSFER_SINGLE_EVENT,
+            from_block=first_stage_tx_receipt.block_number,
+            to_block=first_stage_tx_receipt.block_number,
+        )
+
+        self.assertEqual(len(first_stage_events), first_stage_path_2_count)
+
+        for i, erc1155_transfer_event in enumerate(first_stage_events):
+            self.assertEqual(erc1155_transfer_event["args"]["from"], ZERO_ADDRESS)
+            self.assertEqual(erc1155_transfer_event["args"]["to"], self.player.address)
+            self.assertEqual(erc1155_transfer_event["args"]["id"], self.reward_pool_id)
+            self.assertEqual(
+                erc1155_transfer_event["args"]["value"],
+                (2**1) * (3**2),
```

Comment:
Did we write this together?

It is clever, but it makes it hard to understand what's happening in the test. What do you think @kellan-simiotics ? Should we do it in a more dull but readable way?